snapshotter/devmapper/pool_device.go

```diff
@@ -0,0 +1,289 @@
+	}, nil
+}
+
+func (p *PoolDevice) CreateThinDevice(deviceName string, virtualSizeBytes uint64) (int, error) {
+	mutex.Lock()
+	defer mutex.Unlock()
+
+	if _, ok := p.devices[deviceName]; ok {
+		return 0, errors.Errorf("device with name '%s' already created", deviceName)
+	}
+
+	// Create device, retry if device id is taken
+	deviceID, err := p.tryAcquireDeviceID(func(thinDeviceID int) error {
+		return devicemapper.CreateDevice(p.poolName, thinDeviceID)
+	})
+
+	if err != nil {
+		return 0, errors.Wrap(err, "failed to create thin device")
+	}
+
+	p.devices[deviceName] = deviceID
+
+	devicePath := p.GetDevicePath(p.poolName)
+	if err := devicemapper.ActivateDevice(devicePath, deviceName, deviceID, virtualSizeBytes); err != nil {
+		return 0, errors.Wrap(err, "failed to activate thin device")
+	}
+
+	return deviceID, nil
+}
+
+func (p *PoolDevice) CreateSnapshotDevice(deviceName string, snapshotName string, virtualSizeBytes uint64) (int, error) {
+	mutex.Lock()
+	defer mutex.Unlock()
+
+	deviceID, ok := p.devices[deviceName]
+	if !ok {
+		return 0, errors.Errorf("device '%s' not found", deviceName)
+	}
+
+	if _, ok := p.devices[snapshotName]; ok {
+		return 0, errors.Errorf("snapshot with name '%s' already exists", snapshotName)
+	}
+
+	// Send 'create_snap' message to pool-device
+	devicePoolPath := p.GetDevicePath(p.poolName)
+	thinDevicePath := p.GetDevicePath(deviceName)
+	snapshotDeviceID, err := p.tryAcquireDeviceID(func(snapshotDeviceID int) error {
+		return devicemapper.CreateSnapDevice(devicePoolPath, snapshotDeviceID, thinDevicePath, deviceID)
+	})
+
+	if err != nil {
+		return 0, errors.Wrap(err, "failed to create snapshot")
+	}
+
+	// Activate snapshot
+	if err := devicemapper.ActivateDevice(devicePoolPath, snapshotName, snapshotDeviceID, virtualSizeBytes); err != nil {
+		return 0, errors.Wrap(err, "failed to activate snapshot device")
+	}
+
+	p.devices[snapshotName] = snapshotDeviceID
+	return snapshotDeviceID, nil
+}
+
+func (p *PoolDevice) RemoveDevice(deviceName string) error {
+	mutex.Lock()
+	defer mutex.Unlock()
+
+	return p.removeDevice(deviceName, true)
+}
+
+func (p *PoolDevice) GetDevicePath(deviceName string) string {
+	return getDevMapperPath(deviceName)
+}
+
+func (p *PoolDevice) Close(ctx context.Context, removeDevices, removePool bool) error {
+	mutex.Lock()
+	defer mutex.Unlock()
+
+	var result *multierror.Error
+
+	// Clean thin devices
+	if removeDevices {
+		for name, id := range p.devices {
+			if err := p.removeDevice(name, false); err != nil {
+				log.G(ctx).WithError(err).Errorf("failed to remove device '%s' (id: %d)", name, id)
+				result = multierror.Append(result, err)
+			}
+		}
+	}
+
+	// Remove thin-pool
+	if removePool {
+		if err := p.removePool(ctx); err != nil {
+			log.G(ctx).WithError(err).Errorf("failed to remove thin-pool '%s'", p.poolName)
+			result = multierror.Append(result, err)
+		}
+	}
+
+	return result.ErrorOrNil()
+}
+
+func (p *PoolDevice) getNextDeviceID() int {
```

Comment:
nit: I like putting "unsafe" in the name somewhere for functions like this that expect a mutex to be held (or some other precondition) by the caller

Reply:
Let's leave it as is. Device id is going to be thrown away in subsequent PR in favor of separate pool metadata manager.